Hansard data enrichment utils

shabrf opened the pull request:
Closes https://github.com/nestauk/discovery_utils/issues/45

## Overview
This PR introduces classes for enriching Hansard debate data. It includes functions for data retrieval, processing, and enrichment, which will be used by the Airflow/SkyPilot pipeline implemented in the `discovery-airflow` repository.

#### Changes
- Add `getters/hansard.py` for retrieving Hansard data from S3
- Add `enrichment/hansard.py` with functions for processing and enriching Hansard debates
- Modified `utils/keywords.py` to parametrise the splitting of text into sentences before labelling.

## Implementation Details
- `getter.py`:
  - Includes `HansardGetter` class for fetching data from S3
  - Handles retrieval of existing processed Hansard debates, unprocessed debate XML files, keyword label store, and ML label store
- `enrichment.py`:
  - Contains `HansardProcessor` class for processing raw XML files into a dataframe. This processes the `people.json` metadata file, extracts and cleans debate text from the xmls, and links debates to their respective MP and parties.
  - Contains `HansardEnricher` class for enriching debate data. This Includes both keyword-based and ML-based labelling and handles the appending and uploading of the output label stores.
  - When run, carries out the core enrichment process as outlined below.

#### Data Processing Pipeline
1. Fetch raw data from S3
2. Fetch label stores from S3
3. Fetch processed debates parquet from S3
4. Parse XML files and extract relevant information
5. Append processed debates parquet with newly processed debates
6. Apply keyword-based labelling
7. Apply ML-based labelling
8. Append label stores with labelling outputs
9. Upload results to S3

## Instructions for Reviewer

- Data Retrieval (getter.py):
   - Verify that the `HansardGetter` class correctly interfaces with S3 for all required data (debates, labels, XML files).
   - Check error handling for S3 operations, ensuring failures are handles if files are missing or inaccessible.
- Data Enrichment (enrichment.py):
   - Verify that XML parsing logic in `process_debate_file` correctly extracts all required fields.
   - Review the `HansardEnricher` class for logical flow of data processing steps.
   - Check that text cleaning functions preserve important information while removing noise.
- Labeling Logic:
   - Examine the keyword-based labelling function for efficiency and accuracy.
   - Review the ML model integration and prediction process.
   - Ensure that the labelling process can handle various edge cases (e.g. empty text, unusual characters).
- Performance Considerations:
   - Look for any obvious performance bottlenecks, especially in loops or data transformation steps.
   - Consider if any parts of the code could benefit from parallelisation or vectorisation.
- Error Handling and Logging:
   - Ensure that appropriate error handling is in place for all major operations.
   - Check that logging is implemented to capture important events and errors.
- S3 Interaction:
   - Review the S3 interaction code for security and efficiency.

## Related PRs
https://github.com/nestauk/discovery_airflow/pull/10

shabrf committed:
hansard data enrichment utils

shabrf committed:
refactor core enrichment script from airflow to utils

beingkk commented on discovery_utils/getters/hansard.py:
```diff
@@ -0,0 +1,145 @@
+
+# Configure logging
+logging.basicConfig(level=logging.INFO, format="%(asctime)s - %(name)s - %(levelname)s - %(message)s")
+logger = logging.getLogger(__name__)
```
I suppose we would want to use similar logging config elsewhere. Are there any benefits of copying this to other utils, or shall we move this somewhere else and import here? (eg set up an __init__.py in discovery_utils?)

shabrf replied:
I've moved this to init.py

beingkk commented on discovery_utils/getters/hansard.py:
```diff
@@ -0,0 +1,145 @@
+            logger.error(f"Error downloading CSV file {key}: {str(e)}")
+            raise
+
+    def get_latest_debates(self):
```
Would be good to add a dosctring explaining what's the definition of latest, thank you!

For example: "Download any new XML debate files since the last sync to the "tmp" folder"

beingkk commented on discovery_utils/getters/hansard.py:
```diff
@@ -0,0 +1,145 @@
+        S3 = boto3.client("s3", aws_access_key_id, aws_secret_access_key)
+        return S3
+
+    def get_parquet(self):
```
Would be useful to add a docstring saying that this downloads the processed Hansard debates table

beingkk commented on discovery_utils/enrichment/hansard.py:
```diff
@@ -0,0 +1,348 @@
+            logger.info(f"Successfully appended and uploaded CSV file: {key}")
+        except Exception as e:
+            logger.error(f"Error appending and uploading CSV file {key}: {str(e)}")
+            raise
```
I think it's fine to have these here now - but in principle these seem quite general purpose, so we probably can reuse them later on.

beingkk commented on discovery_utils/enrichment/hansard.py:
```diff
@@ -0,0 +1,348 @@
+
+    def process_new_xml(self, xml_files):
+        logger.info(f"Processing {len(xml_files)} new XML files")
+        return pd.DataFrame(columns=["id", "text"])
```
Is this function used?

beingkk reviewed (changes requested):
Hey @shabrf thanks so much for this, it's shaping up very well!

I have added a few minor comments in the code, and a couple bigger points/bugs that would be great to address before we merge. Thank you!

Edit: I added `tmp` to gitignore and commited

## Data sync bug

Even though I ran get_all_debates(), when I run get_latest_debates() it finds 18 more new files to sync and downloads them again

<img width="793" alt="Screenshot 2024-09-17 at 17 55 15" src="https://github.com/user-attachments/assets/f7efbecf-ae35-45dd-81ee-48668cb0e12e">

## Data and label duplication

One error case I came across is that when appending data, we don't check if the data to be appended already exists in the original. So, if there are some bugs with syncing and we do an incremental update with data that was already processed, we might create duplicates in the enriched parquet file. I think that's why I'm seeing some duplicates already (see below). The same might happen also for label stores. Could we do a quick fix deduplicated the final table after incremental update, or perhaps there's a smarter solution?

<img width="867" alt="Screenshot 2024-09-17 at 17 30 32" src="https://github.com/user-attachments/assets/785076c7-5219-48ea-bcb6-c1c3a299b13d">


## Inconsistent ML label store

Another bug is that ML label store has both `speech_id` and `id` fields, and for some rows we have `speech_id` and other data, while for some rows we have only `id` and `text` fields. Not sure what happened there, but we should harmonise it - I think it would make sense to only have `id` and label-related fields, and perhaps also `text` that was used to infer the labels (unless the text field blows up the file size, but looks like it's OK for now)

<img width="986" alt="Screenshot 2024-09-17 at 17 32 00" src="https://github.com/user-attachments/assets/301cd017-c3e3-4372-bd98-a6d5eaffd45f">

Commit:
gitignoring tmp

beingkk commented on discovery_utils/getters/hansard.py:
```diff
@@ -0,0 +1,145 @@
+        self.s3_prefix = "data/policy_scanning_data/"
+        logger.info(f"Initialized HansardGetter with bucket: {self.bucket}")
+
+    def s3_client(aws_access_key_id: str = AWS_ACCESS_KEY, aws_secret_access_key: str = AWS_SECRET_KEY) -> BaseClient:
+        """Initialize S3 client"""
+        S3 = boto3.client("s3", aws_access_key_id, aws_secret_access_key)
+        return S3
```
Hmm, a bit confused if you're using this function or s3_client function from s3 module imported above to initialise?

shabrf committed:
Docstrings, fix sync and duplication bugs, add reset mode

shabrf committed:
fix dependency conflict

shabrf commented:
Thanks so much for your review @beingkk. I've pushed changes which address your comments, here's a summary of the modifications:
- Added cli arguments to define pipeline mode
- Added a `reset` operational mode to the pipeline to delete the labelstores and run a full refresh
- Testing mode now only enriches 10 debate files and doesn't write to S3
- Remove duplicate entries before uploading to S3
- Standardised columns in label stores to 'id', 'text', and label columns
- Added full data sync check in get_latest_debates to avoid redundant downloads
- Improved error handling
- Added docstrings throughout and additional logging

beingkk commented:
Thank you Shabeer, will check it out!

beingkk commented on discovery_utils/enrichment/hansard.py:
```diff
@@ -1,348 +1,478 @@
-    run_hansard_enrichment()
+    # Parse pipeline mode, default is incremental
+    parser = argparse.ArgumentParser(description="Run Hansard enrichment pipeline")
+    parser.add_argument("--full", action="store_true", help="Run full enrichment instead of incremental")
+    parser.add_argument("--test", action="store_true", help="Use a subset of the data for local testing")
+    parser.add_argument("--reset", action="store_true", help="Reset labelstores and run full enrichment")
```
Great to have these options! What's the difference between "full" an "reset" - is it that if you run "full" it will append the labels to the label store instead of overwriting? 

But I guess if you're now deduplicating the label store, then it will delete the new labels anyways? So, do we need both 'full' and 'reset' options or maybe just 'full' is enough? Let me know though if I'm confused!

shabrf replied:
Good point, I added the reset mode for development purposes to explicitly delete the label stores but on a full run they are overwritten anyway so having both modes is indeed redundant.

beingkk commented on discovery_utils/enrichment/hansard.py:
```diff
@@ -1,348 +1,478 @@
-        self.s3_prefix = "data/policy_scanning_data/"
-        self.bucket = "discovery-iss"
+        Args:
+            bucket (str): The name of the S3 bucket.
+            prefix (str): The prefix for all S3 keys.
```
I guess don't need these two lines anymore?

beingkk reviewed (approved):
Thank you very much @shabrf for the fixes, I think it's good to merge - just had one clarifying question about the "full" vs "reset" options.

shabrf committed:
remove redundant reset pipeline mode

beingkk committed:
Merge branch 'dev' into 40-hansard-enrichment

beingkk commented:
Awesome, merged into dev, thank you! 🙏
